File: src-tauri/src/commands/milestones.rs
// Milestone CRUD commands

use crate::commands::repositories::MilestoneRepository;
use crate::log_command;
use crate::models::Milestone;
use crate::AppState;

#[tauri::command]
pub fn get_milestone(state: tauri::State<AppState>, id: String) -> Result<Milestone, String> {
    log_command!("get_milestone", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        MilestoneRepository::get_by_id_with_progress(&conn, &id)
    })
}

#[tauri::command]
pub fn get_milestones(state: tauri::State<AppState>) -> Result<Vec<Milestone>, String> {
    log_command!("get_milestones", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        MilestoneRepository::get_all_with_progress(&conn)
    })
}

#[tauri::command]
pub fn create_milestone(
    state: tauri::State<AppState>,
    title: String,
    target_date: Option<String>,
    biz_type: Option<String>,
    biz_id: Option<String>,
) -> Result<Milestone, String> {
    log_command!("create_milestone", {
        // Validate: if biz_type is set, biz_id must also be set (and vice versa)
        let has_biz = biz_type.is_some() || biz_id.is_some();
        if has_biz && (biz_type.is_none() || biz_id.is_none()) {
            return Err("Both biz_type and biz_id must be set together, or neither".to_string());
        }

        let conn = state.db.lock().map_err(|e| e.to_string())?;

        let id = uuid::Uuid::new_v4().to_string();
        MilestoneRepository::create(
            &conn,
            &id,
            &title,
            target_date.as_deref(),
            biz_type.as_deref(),
            biz_id.as_deref(),
        )
    })
}

#[tauri::command]
pub fn update_milestone(
    state: tauri::State<AppState>,
    id: String,
    title: Option<String>,
    target_date: Option<String>,
    biz_type: Option<String>,
    biz_id: Option<String>,
    status: Option<String>,
) -> Result<Milestone, String> {
    log_command!("update_milestone", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        MilestoneRepository::update(
            &conn,
            &id,
            title.as_deref(),
            target_date.as_deref(),
            biz_type.as_deref(),
            biz_id.as_deref(),
            status.as_deref(),
        )
    })
}

#[tauri::command]
pub fn delete_milestone(state: tauri::State<AppState>, id: String) -> Result<(), String> {
    log_command!("delete_milestone", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        MilestoneRepository::delete(&conn, &id)
    })
}

File: src/lib/api/milestones.ts
/**
 * Milestone APIs
 *
 * API functions for Milestone entity.
 */

import type {
  Milestone,
  CreateMilestoneParams,
  UpdateMilestoneParams,
} from "@/lib/types";
import { withTauriError, withTauriFallback } from "./utils";

export async function getMilestone(id: string): Promise<Milestone> {
  return withTauriError("get milestones", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<Milestone>("get_milestone", { id });
  });
}

export async function getMilestones(): Promise<Milestone[]> {
  return withTauriFallback(
    "milestones",
    async () => {
      const { invoke } = await import("@tauri-apps/api/core");
      return invoke<Milestone[]>("get_milestones");
    },
    [],
  );
}

export async function getMilestonesByTarget(
  targetId: string,
): Promise<Milestone[]> {
  return withTauriFallback(
    "milestones by target",
    async () => {
      const { invoke } = await import("@tauri-apps/api/core");
      return invoke<Milestone[]>("get_milestones_by_target", {
        target_id: targetId,
      });
    },
    [],
  );
}

export async function createMilestone(
  data: CreateMilestoneParams,
): Promise<Milestone> {
  return withTauriError("create milestones", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<Milestone>("create_milestone", {
      title: data.title,
      target_date: data.target_date || null,
      biz_type: data.biz_type || null,
      biz_id: data.biz_id || null,
    });
  });
}

export async function updateMilestone(
  id: string,
  data: UpdateMilestoneParams,
): Promise<Milestone> {
  return withTauriError("update milestones", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<Milestone>("update_milestone", {
      id,
      title: data.title,
      target_date: data.target_date,
      biz_type: data.biz_type,
      biz_id: data.biz_id,
      status: data.status,
    });
  });
}

export async function deleteMilestone(id: string): Promise<void> {
  return withTauriError("delete milestones", async () => {
    const { invoke } = await import("@tauri-apps/api/core");
    return invoke<void>("delete_milestone", { id });
  });
}
